### ISSUE TITLE
Callbacks for consumer.stop()

### ISSUE THREAD

**Issue by owais** [bbc/sqs-consumer#15]
I'm running this as a long running process and using some external process to manage it (upstart/supervisor/etc). I don't want to forcibly kill the process as it could be in the middle of processing an event. I was hoping to pass a callback to `consumer.stop()` which would be called when the consumer stop and all existing messages are processed. The idea is to do something like this,

```
consumer.start();

process.on("SIGTERM", function() {
     consumer.stop(function(){
         process.exit(0);
     });
});
```

**Comment by robinjmurphy** [comment]
This is an interesting one. It wouldn't be straight forward to implement currently as the 'stop' method is just a flag to temporarily stop polling. It's not really a 'shutdown' function.

In theory, your app should be able to cope with the process being killed. There's a chance that a message will have been 'processed' but not deleted, but that means it will simple be left on the queue and reprocessed when your app starts up again.

Is there a specific use case/problem that is causing you issues at the moment?

**Comment by owais** [comment]
> In theory, your app should be able to cope with the process being killed. There's a chance that a message will have been 'processed' but not deleted, but that means it will simply be left on the queue and reprocessed when your app starts up again. Given the one-or-more delivery semantics of SQS, your app should be able to cope with the same message being processed twice.

It is not always OK to process an event twice. For example, we wouldn't want to send out a duplicate email. I can work around this issue by adding a local result store/cache to mark events as done and ignore them if they arrive again but that sounds like a big overkill as it should be SQS doing that, not me.

**Comment by hjerling** [comment]
Hi @owais, 
The [at-least-once delivery](http://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/DistributedQueues.html) is a property of AWS SQS and nothing we can do anything about. All apps that use AWS SQS need to handle this separately as an only-once delivery is never guaranteed by SQS. So even if your app runs perfectly fine and never ever gets forcibly killed you might still end up with the same message being delivered more than once.

**Comment by owais** [comment]
@hjerling That makes complete sense but IMHO, any such system (web worker, task worker,etc) should support "graceful reloads". I think it is counter-intuitive when the consumer quits without waiting for a callback to be fired first.

@robinjmurphy @hjerling 
I can work on this in my limited free time if it sounds something like you'd be willing to accept a feature like this.

**Comment by robinjmurphy** [comment]
@owais By all means share an example. We're just used to building systems where you shouldn't assume that a `.stop()` has been called - your app/instance can die at any time without politely giving you a chance to clean up.

**Comment by robinjmurphy** [comment]
Fixed in #41
